src/kvstore/raftex/RaftexService.cpp

```diff
@@ -21,57 +21,97 @@ std::shared_ptr<RaftexService> RaftexService::createService(
         std::shared_ptr<folly::IOThreadPoolExecutor> pool,
         uint16_t port) {
     auto svc = std::shared_ptr<RaftexService>(new RaftexService());
+    CHECK(svc != nullptr) << "Failed to create a raft service";
 
     svc->server_ = std::make_unique<apache::thrift::ThriftServer>();
     CHECK(svc->server_ != nullptr) << "Failed to create a thrift server";
-
     svc->server_->setInterface(svc);
-    svc->server_->setPort(port);
+
+    svc->initThriftServer(pool, port);
+    return svc;
+}
+
+
+RaftexService::~RaftexService() {
+}
+
+
+bool RaftexService::start() {
+    serverThread_.reset(new std::thread([&] {
+        serve();
+    }));
+
+    waitUntilReady();
+    if (status_ != STATUS_RUNNING) {
+        waitUntilStop();
+        server_.reset();
+        return false;
+    }
+
+    return true;
+}
+
+
+void RaftexService::initThriftServer(std::shared_ptr<folly::IOThreadPoolExecutor> pool, 
+                                     uint16_t port) {
+    LOG(INFO) << "Init thrift server for raft service.";
+    server_->setPort(port);
     if (pool != nullptr) {
-        svc->server_->setIOThreadPool(pool);
+        server_->setIOThreadPool(pool);
     }
+}
 
-    svc->serverThread_.reset(new std::thread([svc] {
-        LOG(INFO) << "Starting the Raftex Service";
 
-        svc->server_->setup();
-        svc->serverPort_ = svc->server_->getAddress().getPort();
-        SCOPE_EXIT {
-            svc->server_->cleanUp();
-        };
+bool RaftexService::setup() {
+    LOG(INFO) << "Starting the Raftex Service";
 
-        {
-            std::lock_guard<std::mutex> g(svc->readyMutex_);
-            svc->ready_ = true;
-        }
-        svc->readyCV_.notify_all();
+    try {
+        server_->setup();
```

Comment:
Is the step make the whole process crashed if port is used?

Reply:
yes, server_->setup() will call “AsyncServerSocket::bindSocket”, and this function will throw an exception
if (fsp::bind(fd, saddr, address.getActualSize()) != 0) {
    if (!isExistingSocket) {
      closeNoInt(fd);
    }
    folly::throwSystemError(
        errno, "failed to bind to async server socket: " + address.describe());
  }

then I can see it in the nebula-storaged.INFO:
E0716 14:20:59.741447  2314 ThriftServer.cpp:440] Got an exception while setting up the server: 98failed to bind to async server socket: [::]:18002: Address already in use

Reply:
So we could handle it in a more simple way.  Just move server_->setup out from the background thread,  and return error if failed.  In waitUntilReady,  after starting up the event loop in bg thread,  we could pend there to check the loop status before returned.  

PS.  You could take the base/test/ServerContext.h as your reference.

Reply:
I have think about your ideal deeply with see the source of socket setup, and you are right, I will fix it, thanks very much.

Reply:
I check the function setup again, the function will init event base by “serveEventBase_ = eventBaseManager_->getEventBase();”, and this is thread local variable, so I can not call set up in main thread.
and I will use a atomic_int to replace the “status_ and cv”, thx.